tokio-trace/src/span.rs

```diff
@@ -341,14 +362,78 @@ impl Span {
     /// [`Subscriber::enter`]: ../subscriber/trait.Subscriber.html#method.enter
     /// [`Subscriber::exit`]: ../subscriber/trait.Subscriber.html#method.exit
     /// [`Id`]: ../struct.Id.html
+    /// [`enter_unscoped`]: #method.enter_unscoped
     pub fn enter<'a>(&'a self) -> Entered<'a> {
-        if let Some(ref inner) = self.inner.as_ref() {
-            inner.subscriber.enter(&inner.id);
-        }
-        self.log(format_args!("-> {}", self.meta.name));
+        self.enter_inner();
         Entered { span: self }
     }
 
+    /// Enters the span, moving `self` into a guard which will exit the span
+    /// when dropped.
+    ///
+    /// Unlike [`enter`], the guard returned by `enter_unscoped` _moves_ the
+    /// span handle, and may therefore outlive the scope in which this method is
+    /// called.
+    ///
+    /// If this span is enabled by the current subscriber, then this function will
+    /// call [`Subscriber::enter`] with the span's [`Id`], and dropping the guard
+    /// will call [`Subscriber::exit`]. If the span is disabled, this does
+    /// nothing.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #[macro_use] extern crate tokio_trace;
+    /// # use tokio_trace::Level;
+    /// # fn main() {
+    /// let entered = span!(Level::INFO, "my_span").enter_unscoped();
+    /// # }
+    /// ```
+    ///
+    /// Returning an entered span guard as part of a struct:
+    /// ```
+    /// # #[macro_use] extern crate tokio_trace;
+    /// # use tokio_trace::Level;
+    /// # fn main() {
+    /// pub struct Foo {
+    ///     name: &'static str,
+    ///     // ...
+    /// }
+    ///
+    /// pub struct FooWithSpan {
+    ///     enter: tokio_trace::span::EnteredUnscoped,
+    ///     foo: Foo,
+    /// }
+    ///
+    /// impl Foo {
+    ///     pub fn new(name: &'static str) -> Self {
+    ///         Foo {
+    ///             name,
+    ///            // ...
+    ///         }
+    ///     }
+    ///
+    ///     pub fn new_with_span(name: &'static str) -> FooWithSpan {
+    ///         let foo = Foo::new(name);
+    ///         let enter = span!(Level::DEBUG, "foo", name).enter_unscoped();
+    ///         // The span will remain entered until the `FooWithSpan` is dropped.
+    ///         FooWithSpan {
+    ///             foo,
+    ///             enter,
+    ///         }
+    ///     }
+    /// }
+    /// # }
+    /// ```
+    /// [`Subscriber::enter`]: ../subscriber/trait.Subscriber.html#method.enter
+    /// [`Subscriber::exit`]: ../subscriber/trait.Subscriber.html#method.exit
+    /// [`Id`]: ../struct.Id.html
+    /// [`enter`]: #method.enter
+    pub fn enter_unscoped(self) -> EnteredUnscoped {
```

Comment:
I'm open to bikeshedding on the name; `enter_unscoped` seemed more descriptive of the intended _use case_ than `enter_owned`, but the latter is shorter...

Reply:
![Screen Shot 2019-06-18 at 1 02 24 PM](https://user-images.githubusercontent.com/2796466/59722261-19017480-91d8-11e9-9d0e-0400befd6efe.png)
☝️ is the objectively correct name